witness/golang/cmd/witness/internal/witness/witness.go

```diff
@@ -155,75 +154,153 @@ func (w *Witness) Update(ctx context.Context, logID string, chkptRaw []byte, pro
 	}
 	// Check the signatures on the raw checkpoint and parse it
 	// into the log.Checkpoint format.
-	chkpt, err := w.parse(chkptRaw, logID)
+	chkptNew, err := w.parse(chkptRaw, logID)
 	if err != nil {
 		return 0, err
 	}
-	c := Chkpt{Size: chkpt.Size, Raw: chkptRaw}
 	// Get the latest one for the log because we don't want consistency proofs
 	// with respect to older checkpoints.  Bind this all in a transaction to
 	// avoid race conditions when updating the database.
 	tx, err := w.db.BeginTx(ctx, nil)
 	if err != nil {
 		return 0, fmt.Errorf("couldn't create tx: %v", err)
 	}
-	latest, err := w.getLatestChkpt(tx.QueryRow, logID)
+	// Get the latest checkpoint and compact range (if one exists).
+	latestRaw, rangeRaw, err := w.getLatestChkptData(tx.QueryRow, logID)
 	if err != nil {
 		// If there was nothing stored already then treat this new
 		// checkpoint as trust-on-first-use.
 		if status.Code(err) == codes.NotFound {
-			return 0, w.setCheckpoint(tx, logID, c)
+			// If we're using compact ranges then store the input
+			// proof as our initial range, assuming it matches the
+			// input checkpoint.
+			if logInfo.UseCompact {
+				rf := compact.RangeFactory{Hash: logInfo.Hasher.HashChildren}
+				rng, err := rf.NewRange(0, chkptNew.Size, proof)
+				if err != nil {
+					return 0, fmt.Errorf("can't form compact range: %v", err)
+				}
+				if err := w.verifyRangeHash(chkptNew.Hash, rng); err != nil {
+					return 0, fmt.Errorf("input root hash doesn't verify: %v", err)
+				}
+				rangeRaw := []byte(log.Proof(proof).Marshal())
+				return 0, w.setChkptData(tx, logID, chkptRaw, rangeRaw)
+			}
+			return 0, w.setChkptData(tx, logID, chkptRaw, nil)
 		}
 		return 0, fmt.Errorf("Update: %w", err)
 	}
-	p, err := w.parse(latest.Raw, logID)
+	// Parse the raw checkpoint into the log.Checkpoint format.
+	chkptOld, err := w.parse(latestRaw, logID)
 	if err != nil {
-		return 0, err
+		return 0, fmt.Errorf("couldn't parse stored checkpoint: %v", err)
+	}
+	// Parse the compact range if we're using one.
+	var curRange log.Proof
+	if logInfo.UseCompact {
+		if err := curRange.Unmarshal(rangeRaw); err != nil {
+			return 0, fmt.Errorf("couldn't unmarshal proof: %v", err)
+		}
 	}
 	switch {
```

Comment:
There is one dominating `case` in this `switch`. I think it would be more readable if the error "new < old" case and the fast-path "new == old" case were put in the beginning as two `if`s, and then the dominating "new > old" case was left as a remaining unconditional flow (with the benefit of not needing to indent a big chunk of code). You wouldn't need the `default` boilerplate too.